File: leetcode/medium/973-kClosestPointsToOrigin.js
// first pass solution
var kClosest = function(points, k) {
    let dict = {};
    let output = [];
    
    for (let i = 0; i < points.length; i++) {
        let point = points[i];
        dict[i] = point[0] * point[0] + point[1] * point[1];
    }
    
    let arr = Object.entries(dict).sort((a, b) => a[1] - b[1]);
    let i = 0;
    
    while (output.length < k) {
        output.push(points[arr[i][0]]);
        i++;
    }
    return output;
};

File: leetcode/medium/973-kClosestPointsToOrigin.java
class Solution {
    public int[][] kClosest(int[][] points, int k) {
        int[][] output = new int[k][2];
        Queue<int[]> heap = new PriorityQueue<>(
            (a, b) -> (b[0] * b[0] + b[1] * b[1]) - (a[0] * a[0] + a[1] * a[1]));
        
        for (int[] point : points) {
            heap.add(point);
            if (heap.size() > k) heap.poll();
        }
        
        for (int i = k - 1; i >= 0; i--) {
            output[i] = heap.poll();
        }
        
        return output;
    }
}
